### ISSUE TITLE
bug: runs very sporadically.

### ISSUE THREAD

**Issue by sinkr** [MisanthropicBit/decipher.nvim#11]
### Did you check docs and existing issues?

- [X] I have read all the docs
- [X] I have searched the existing issues
- [X] I have searched the existing issues of plugins related to this issue (if relevant)

### Neovim version (nvim -v)

v0.10.0-dev-2947+ge1ca7a7bf-dirty

### Operating system/version

MacOS 14.4.1

### Describe the bug

Single word text selection -> base64, yields error on simple text file..
```
[decipher]: base64: ...e/nvim/lazy/decipher.nvim/lua/decipher/codecs/base64.lua:153: attempt to index global 'bit' (a function value)
```

### Steps To Reproduce

1. Open a empty file,
2. input a simple string,
3. require("decipher").encode_selection_prompt()
4. Select 'base64'

### Expected Behavior

base64 text occurs.

**Comment by MisanthropicBit** [comment]
Thanks for submitting an issue and using the issue template 🙂

You say the bug occurs sporadically and I found that I am using neovim's `bit` library directly for only the base64 codec instead of decipher's [`bits` wrapper](https://github.com/MisanthropicBit/decipher.nvim/blob/master/lua/decipher/bits.lua) module. My guess is that it occurs sporadically because something else is sometimes doing a `require("lua")` somewhere beforehand, and making the `bit` library available, and sometimes not, making it fail.

You can test out the fix on [this branch](https://github.com/MisanthropicBit/decipher.nvim/tree/fix/always-use-bits-wrapper) but I'll merge it regardless since this is a bug even if it doesn't fix your issue.

**Comment by sinkr** [comment]
Thank you for the plugin and the response!  This works better in non-preview mode for sure.  When I try to do a preview, I see the box pop up w/ the correct value and then it disappears in under a second.

I can live without the preview, however.

**Comment by sinkr** [comment]
I figured it out!  `autoclose = false`

**Comment by MisanthropicBit** [comment]
> This works better in non-preview mode for sure.

Great 🙂 thanks for testing it out so quickly.

> I figured it out! `autoclose = false`

The preview window should only be closing if you enter insert mode or move the cursor right after the pop-up window opens. I originally made a fix for the behavior you are describing but specifically for motions since they would move the cursor after doing the encoding/decoding, closing the window immediately. I never had this problem with the selection functions.

* Does the window close itself even if you don't trigger any of these events after opening the window?
* In your repro, I'm assuming that you visually select the text between steps 2 and 3, right?

Doing these steps, I cannot reproduce your issue (running neovim v0.9.5 on OSX 11.7.10):

1. Open a empty file.
2. Input a simple string.
3. Visually select the text.
4. Execute `'<,'>lua require("decipher").encode_selection_prompt({ preview = true })` (works with `'<,'>lua require("decipher").encode_selection('base64')` as well).
5. Select 'base64'.

**Comment by MisanthropicBit** [comment]
FYI, I've released [v1.0.1](https://github.com/MisanthropicBit/decipher.nvim/releases/tag/v1.0.1) that includes the above mentioned fix https://github.com/MisanthropicBit/decipher.nvim/issues/11#issuecomment-2068189885

**Comment by MisanthropicBit** [comment]
@sinkr Did you have a chance to verify that v1.0.1 fixed your issue?

It would also be great if you could check if either [fix/float-autoclose-autocmds](https://github.com/MisanthropicBit/decipher.nvim/tree/fix/float-autoclose-autocmds) or [fix/float-autoclose-autocmds2](https://github.com/MisanthropicBit/decipher.nvim/tree/fix/float-autoclose-autocmds2) fixes your issue with the preview closing with `autoclose = true` 🙂

**Comment by sinkr** [comment]
Sadly no, it doesn't work for me, still.  I've fallen back to using [b64.nvim;](https://github.com/taybart/b64.nvim) I'd love to have the preview/JSON feature, but damned if I can make it work.  It's probably PEBKAC.  Below's my config and a GIF of what I'm seeing.

```
return {
  "https://github.com/MisanthropicBit/decipher.nvim.git",

  init = function()
    require("decipher").setup({

      vim.keymap.set("v", "<leader>1", function()
        require("decipher").encode_selection_prompt({ preview = true })
      end, { desc = "Decipher Encode", noremap = true, silent = true }),

      vim.keymap.set("v", "<leader>2", function()
        require("decipher").decode_selection_prompt({ preview = true })
      end, { desc = "Decipher Decode", noremap = true, silent = true }),

      active_codecs = "all", -- Set all codecs as active and useable
      float = { -- Floating window options
        padding = 0, -- Zero padding (does not apply to title if any)
        border = { -- Floating window border
          { "╭", "FloatBorder" },
          { "─", "FloatBorder" },
          { "╮", "FloatBorder" },
          { "│", "FloatBorder" },
          { "╯", "FloatBorder" },
          { "─", "FloatBorder" },
          { "╰", "FloatBorder" },
          { "│", "FloatBorder" },
        },
        mappings = {
          close = "q", -- Key to press to close the floating window
          apply = "a", -- Key to press to apply the encoding/decoding
          jsonpp = "J", -- Key to prettily format contents as json if possbile
          help = "?", -- Toggle help
        },
        title = true, -- Display a title with the codec name
        title_pos = "left", -- Position of the title
        autoclose = false, -- Autoclose floating window if insert
        -- mode is activated or the cursor is moved
        enter = true, -- Automatically enter the floating window if
        -- opened
        options = {}, -- Options to apply to the floating window contents
      },
    })
  end,
}
```
![blah](https://github.com/MisanthropicBit/decipher.nvim/assets/696374/67b2ec41-c259-476e-8a28-d2a594b19b3c)

**Comment by sinkr** [comment]
So, I did a little more playing around and pretty much gave up on the `selection_prompt` variants. 

What I found is the direction of the selection seems to matter:

To test, I would input text, hit <ESC> v 0 and essentially select from $ to 0; this does not work.
Somewhere along the way, I got fatigued and selected from 0 (in this case ^) to $ and it consistently works.  See attached animation from [fix/float-autoclose-autocmds2](https://github.com/MisanthropicBit/decipher.nvim/tree/fix/float-autoclose-autocmds2):
![blah2](https://github.com/MisanthropicBit/decipher.nvim/assets/696374/a1d31b3d-3952-4a0c-8852-efc42647555a)

**Comment by sinkr** [comment]
Finally, the **FIRST** time I try to run the plugin it ***ALWAYS*** fails:
![blah3](https://github.com/MisanthropicBit/decipher.nvim/assets/696374/c3129a39-f75d-4805-b6e7-44cd8310d0d7)

**Comment by MisanthropicBit** [comment]
> Did you have a chance to verify that v1.0.1 fixed your issue?

Here, I was referring to your original issue regarding `'attempt to index global 'bit'` but it looks from the recordings (very helpful btw, thanks) that you can use the base64 codec without that issue now.

Your config looks fine to me so I'm not sure this is PEBKAC. Had a good laugh though, never heard that term before 🙂

I've tried to reproduce your issue but without any luck. I've tried with visual mode character-, line-, and block-wise. When I have some more time, I'll try and install your version of neovim (looks like some version of nightly) and see if I can reproduce it. Perhaps something changed regarding visual selection.

From the error message, it looks like neovim is complaining that you it cannot get text when the starting column is greater than the end column are reversed but the plugin [already takes this into account](https://github.com/MisanthropicBit/decipher.nvim/blob/master/lua/decipher/selection.lua#L96-L99) for block visual mode where this is applicable.

You could also try to run `nvim -u minimal_init.lua` to make sure that no other plugin is interfering where `minimal_init.lua` is a file like:

```lua
vim.opt.rtp:append("<path_to_your_installation_of_decipher>")
```

Also thanks for making me aware of `b64.nvim`. I didn't know about that plugin.

**Comment by sinkr** [comment]
Yep, it works out of the box w/ the minimal Lua, so I guess it's something else in my setup interfering.  I appreciate the help, however, because while I've been using Vim for several decades, I never decided to really enmesh myself in it and I'm trying to learn the ins and outs, especially since Lua came on the scene.

Just for my own edification, what would be the next steps be to debug this issue?  I find that LazyVim and Noice seemingly hide a lot of error messages, but I feel like there should be a debugger somewhere, maybe Vimspector?

Thanks for your time and the plugin!

**Comment by MisanthropicBit** [comment]
Great! 🙂 Glad to hear it's working.

I've had some time to test this out now with your exact settings and I managed to trigger *almost* the same error.

```
E5108: Error executing lua: .../projects/vim/decipher.nvim/lua/decipher/selection.lua:224: Invalid 'end_col': out of range
stack traceback:
        [C]: in function 'set_visual_text'
        .../projects/vim/decipher.nvim/lua/decipher/selection.lua:224: in function 'set_text_from_selection'
        .../projects/vim/decipher.nvim/lua/decipher/ui/float.lua:354: in function 'apply_codec'
        .../projects/vim/decipher.nvim/lua/decipher/ui/float.lua:301: in function <.../projects/vim/decipher.nvim/lua/decipher/ui/float.lua:300>
```

With your settings, the steps to reproduce are:

1. Enter some text.
2. Do `0v$` or `0<c-v>$` in normal mode.
3. Press `<leader>1`.
4. Select a codec.
5. Press `a` to apply the codec.

If I instead do `0g_` in step 2 (to select to the last non-blank character of the line), it works fine. I couldn't reproduce your error with `start_col` by doing `$v0` or `$v^` with any of the other visual modes as well. I couldn't reproduce your startup issue either (with 0.9.5 or nightly).

```
NVIM v0.10.0-dev-3085+gb024643ca
Build type: RelWithDebInfo
LuaJIT 2.1.1713484068
```

Tested on both master, [fix/float-autoclose-autocmds](https://github.com/MisanthropicBit/decipher.nvim/tree/fix/float-autoclose-autocmds), and [fix/float-autoclose-autocmds2](https://github.com/MisanthropicBit/decipher.nvim/tree/fix/float-autoclose-autocmds2). I'll fix the issue above and then I'll mark this issue as resolved.

> because while I've been using Vim for several decades, I never decided to really enmesh myself in it and I'm trying to learn 
> the ins and outs, especially since Lua came on the scene.

I myself come from a decade of using vim so I completely understand. Perhaps [this](https://neovim.io/doc/user/nvim.html#nvim-from-vim) is helpful? Especially the link to `vim-differences` and `nvim-features` are nice. Also `:h luaref-intro`.

> Just for my own edification, what would be the next steps be to debug this issue? I find that LazyVim and Noice seemingly 
> hide a lot of error messages, but I feel like there should be a debugger somewhere, maybe Vimspector?

I have yet to see a great solution to this problem but I've had a similar problem before where I ended up doing a manual "git bisect" by disabling half my plugins, checking if the problem remained, and then gradually honing in on the problem. You may perhaps already have an idea of what types of plugins might be causing the issue. Disabling noice might also be a good start if you think it's hiding error messages.

Using Vimspector would probably be a good thing to try since it can attach to a running process so you don't have to set breakpoints and step through a lot of code. You could also try to see if there are other autocmds loaded that might interfere, e.g. doing `verbose autocmd WinEnter`.

You could also try to downgrade to the latest stable release (v0.9.5 as of this writing) to see if your particular version of nightly is the problem.

**Comment by AntonMaxen** [comment]
Is this fixed yet?

**Comment by MisanthropicBit** [comment]
What specific issue are you referring to here? Multiple ones came up during the discussions above.

The selection issue should be partially fixed by #18. You are welcome to try out the branch and see if it works for you. I'll try to make time to test it properly and release it but life keeps getting in the way 🙂 OP discovered that this issue was caused by their setup. You can try to use the `minimal_init.lua` approach outlined above to see if that is the case for you as well if #18 doesn't help.
